modules/tinymce/src/core/main/ts/mode/Mode.ts

```diff
@@ -1,16 +1,25 @@
-import { Arr, Cell, Obj } from '@ephox/katamari';
+import { Arr, Cell, Obj, Type } from '@ephox/katamari';
 
 import Editor from '../api/Editor';
 import * as Events from '../api/Events';
-import { EditorModeApi } from '../api/Mode';
+import { EditorModeApi, ReadOnlyProperty } from '../api/Mode';
 import { toggleReadOnly } from './Readonly';
 
 const defaultModes = [ 'design', 'readonly' ];
 
-const switchToMode = (editor: Editor, activeMode: Cell<string>, availableModes: Record<string, EditorModeApi>, mode: string) => {
-  const oldMode = availableModes[activeMode.get()];
+const switchToMode = (editor: Editor, activeMode: Cell<[string, EditorModeApi]>, availableModes: Record<string, EditorModeApi>, mode: string) => {
+  const [ , oldMode ] = activeMode.get();
   const newMode = availableModes[mode];
 
+  // if deactivate fails, hope nothing bad happened and abort
```

Comment:
This is actually the opposite of what I intended; activate and deactivate are "setup and teardown in reaction to the mode changing", not control mechanisms.

I'm not opposed to the change, exceptions in deactivate should not be normal, so it doesn't really matter what we do.